File: companion-android/app/src/main/java/com/axm/companion/protocol/Protocol.kt
package com.axm.companion.protocol

import org.json.JSONObject

/**
 * The wire contract with A-X-M. This is the Kotlin side of
 * src/main/companion/protocol.ts in the A-X-M repository; the two are kept in step
 * by diffing those two files, so any change here needs the same change there.
 *
 * The rules that matter:
 *
 * - Every frame is one JSON object carrying `protocol`, `type` and `payload`. A
 *   version we do not know is reported and the socket closed, rather than guessing
 *   at a payload whose shape may have moved.
 * - Nothing the phone can send names a file, a path or a command. Every outbound
 *   message picks from a fixed set; the host decides what those mean. That is what
 *   makes an open port on the host acceptable.
 */
object Protocol {
    const val VERSION = 1
    const val DISCOVERY_PORT = 47820
    const val CONTROL_PORT = 47821
    const val DISCOVERY_MAGIC = "AXM-COMPANION-DISCOVER"

    // Types the host accepts before pairing. Anything else is refused and the
    // socket dropped, so there is no point sending it.
    const val DEVICE_HELLO = "device.hello"
    const val DEVICE_WELCOME = "device.welcome"
    const val DEVICE_GOODBYE = "device.goodbye"
    const val PAIR_SUBMIT = "pair.submit"
    const val PAIR_RESULT = "pair.result"
    const val LINK_PING = "link.ping"
    const val LINK_PONG = "link.pong"
    const val LINK_ERROR = "link.error"

    const val COMPANION_MODE = "companion.mode"
    const val XMB_INPUT = "xmb.input"
    const val MEDIA_COMMAND = "media.command"
    const val MEDIA_STATE = "media.state"
    const val POINTER_INPUT = "pointer.input"
    const val TOYBOX_SCAN = "toybox.scan"
    const val TOYBOX_METADATA = "toybox.metadata"
    const val GHOST_STATE = "ghost.state"
    const val GHOST_MESSAGE = "ghost.message"
    const val SETTINGS_STATE = "settings.state"
    const val SETTINGS_SET = "settings.set"
    const val KEYBOARD_SHOW = "keyboard.show"
    const val KEYBOARD_HIDE = "keyboard.hide"
    const val KEYBOARD_INPUT = "keyboard.input"
    const val MUSIC_BROWSE = "music.browse"
    const val MUSIC_LISTING = "music.listing"
    const val MUSIC_PLAY = "music.play"
    const val SAVES_LIST = "saves.list"
    const val SAVES_FILE = "saves.file"
    const val SAVES_REQUEST = "saves.request"
    const val SAVES_PUSH = "saves.push"
    const val SAVES_CHEATS = "saves.cheats"
    const val SAVES_PATCHES = "saves.patches"
    const val SAVES_APPLY = "saves.apply"
    const val SAVES_RESTORE = "saves.restore"
    const val SAVES_RESULT = "saves.result"

    // Experimental: the Device column in A-X-M - this phone's calls, texts and
    // contacts. The host asks; the phone answers only for what the user shares.
    const val PHONE_STATE = "phone.state"
    const val PHONE_REQUEST = "phone.request"
    const val PHONE_CONTACTS = "phone.contacts"
    const val PHONE_THREADS = "phone.threads"
    const val PHONE_MESSAGES = "phone.messages"
    const val PHONE_CALLS = "phone.calls"
    const val PHONE_DIAL = "phone.dial"
    const val PHONE_ANSWER = "phone.answer"
    const val PHONE_HANGUP = "phone.hangup"
    const val PHONE_CALL_STATE = "phone.callState"
    const val PHONE_SEND_SMS = "phone.sendSms"
    const val PHONE_SMS_RESULT = "phone.smsResult"
    const val PHONE_INCOMING_SMS = "phone.incomingSms"

    fun frame(type: String, payload: JSONObject): String =
        JSONObject()
            .put("protocol", VERSION)
            .put("type", type)
            .put("payload", payload)
            .toString()

    /** Never throws: this parses bytes off the network, where anything can arrive. */
    fun parse(raw: String): Frame? = try {
        val o = JSONObject(raw)
        val type = o.optString("type", "")
        val payload = o.optJSONObject("payload")
        if (type.isEmpty() || payload == null) null
        else Frame(o.optInt("protocol", -1), type, payload)
    } catch (_: Exception) {
        null
    }
}

data class Frame(val protocol: Int, val type: String, val payload: JSONObject)

/** What this phone can do. The host switches features on from this, never assumes. */
data class Capabilities(
    val touch: Boolean = true,
    val nfc: Boolean = false,
    val foldable: Boolean = false,
    val gyro: Boolean = false,
    val haptics: Boolean = true,
    val videoDecode: Boolean = true,
    val camera: Boolean = false,
    val microphone: Boolean = false,
) {
    fun toJson(): JSONObject = JSONObject()
        .put("touch", touch)
        .put("nfc", nfc)
        .put("foldable", foldable)
        .put("gyro", gyro)
        .put("haptics", haptics)
        .put("videoDecode", videoDecode)
        .put("camera", camera)
        .put("microphone", microphone)
}

/** An A-X-M host that answered a discovery probe. */
data class DiscoveredHost(
    val address: String,
    val controlPort: Int,
    val hostName: String,
    val hostVersion: String,
    val protocol: Int,
)

/** Which screen the phone is showing. The host can ask for a change. */
enum class CompanionMode(val wire: String) {
    IDLE("idle"),
    MEDIA_REMOTE("media_remote"),
    TOUCHPAD("touchpad"),
    XMB_CONTROLLER("xmb_controller"),
    DS_SECOND_SCREEN("ds_second_screen"),
    THREE_DS_SECOND_SCREEN("3ds_second_screen"),
    TOY_BOX("toy_box"),
    GHOST("ghost");

    companion object {
        fun from(wire: String): CompanionMode =
            entries.firstOrNull { it.wire == wire } ?: IDLE
    }
}

/** Menu actions, matching exactly what the A-X-M menu already understands. */
enum class XmbAction(val wire: String) {
    UP("up"), DOWN("down"), LEFT("left"), RIGHT("right"),
    CONFIRM("confirm"), BACK("back"), CONTEXT("context"), GUIDE("guide")
}

enum class MediaCommand(val wire: String) {
    PLAY("play"), PAUSE("pause"), TOGGLE("toggle"), NEXT("next"), PREVIOUS("previous"),
    STOP("stop"), VOLUME("volume"), MUTE("mute"), SHUFFLE("shuffle"),
    REPEAT("repeat"), FAVORITE("favorite"), SEEK("seek"), POSITION("position"), ROUTE("route")
}

/** What A-X-M says is playing. The host is the authority; this only mirrors it. */
data class MediaState(
    val playing: Boolean = false,
    val title: String? = null,
    val artist: String? = null,
    val album: String? = null,
    val artworkUrl: String? = null,
    val positionSeconds: Double = 0.0,
    val durationSeconds: Double = 0.0,
    val volume: Double = 1.0,
    val muted: Boolean = false,
    val streamUrl: String? = null,
    val output: String = "host",
) {
    companion object {
        fun from(p: JSONObject) = MediaState(
            playing = p.optBoolean("playing", false),
            title = p.optString("title").ifEmpty { null },
            artist = p.optString("artist").ifEmpty { null },
            album = p.optString("album").ifEmpty { null },
            artworkUrl = p.optString("artworkUrl").ifEmpty { null },
            positionSeconds = p.optDouble("positionSeconds", 0.0),
            durationSeconds = p.optDouble("durationSeconds", 0.0),
            volume = p.optDouble("volume", 1.0),
            muted = p.optBoolean("muted", false),
            streamUrl = p.optString("streamUrl").ifEmpty { null },
            output = p.optString("output").ifEmpty { "host" },
        )
    }
}

/**
 * A menu setting the host lets this phone change. The host names it, says what it
 * can be and what it is; the phone only ever sends back an id and one of those values.
 */
data class CompanionSetting(
    val id: String,
    val title: String,
    val group: String,
    val kind: String,          // "toggle" or "choice"
    val value: String,         // "on" / "off", or an option id
    val options: List<Pair<String, String>>,   // id to label, for a choice
    val detail: String?,
) {
    companion object {
        fun list(p: JSONObject): List<CompanionSetting> {
            val arr = p.optJSONArray("items") ?: return emptyList()
            return (0 until arr.length()).mapNotNull { i ->
                val o = arr.optJSONObject(i) ?: return@mapNotNull null
                val opts = o.optJSONArray("options")
                CompanionSetting(
                    id = o.optString("id"),
                    title = o.optString("title"),
                    group = o.optString("group", "Settings"),
                    kind = o.optString("kind", "toggle"),
                    value = o.optString("value"),
                    options = if (opts == null) emptyList() else (0 until opts.length()).mapNotNull { j ->
                        opts.optJSONObject(j)?.let { it.optString("id") to it.optString("label") }
                    },
                    detail = o.optString("detail").ifEmpty { null },
                )
            }
        }
    }
}

/** The menu is asking for text; the phone's keyboard can supply it. */
data class KeyboardPrompt(val title: String, val label: String, val value: String, val secret: Boolean) {
    companion object {
        fun from(p: JSONObject) = KeyboardPrompt(
            title = p.optString("title"),
            label = p.optString("label"),
            value = p.optString("value"),
            secret = p.optBoolean("secret", false),
        )
    }
}

/** One folder of the host's music library. Keys are the host's; the phone never sees a path. */
data class MusicListing(val key: String, val name: String, val parent: String?, val entries: List<MusicEntry>) {
    companion object {
        fun from(p: JSONObject): MusicListing {
            val arr = p.optJSONArray("entries")
            return MusicListing(
                key = p.optString("key"),
                name = p.optString("name", "Music"),
                parent = if (p.has("parent")) p.optString("parent") else null,
                entries = if (arr == null) emptyList() else (0 until arr.length()).mapNotNull { i ->
                    arr.optJSONObject(i)?.let { MusicEntry(it.optString("key"), it.optString("name"), it.optString("kind") == "folder") }
                },
            )
        }
    }
}

data class MusicEntry(val key: String, val name: String, val folder: Boolean)

/** The memory card saves the host manages, with a hash per save so the phone knows what it lacks. */
data class SaveListing(val cards: List<SaveCard>, val autoSync: Boolean) {
    companion object {
        fun from(p: JSONObject): SaveListing {
            val arr = p.optJSONArray("cards")
            val cards = if (arr == null) emptyList() else (0 until arr.length()).mapNotNull { i ->
                arr.optJSONObject(i)?.let { c ->
                    val s = c.optJSONArray("saves")
                    SaveCard(
                        id = c.optString("id"), name = c.optString("name"), kind = c.optString("kind"),
                        saves = if (s == null) emptyList() else (0 until s.length()).mapNotNull { j ->
                            s.optJSONObject(j)?.let { SaveEntry(it.optString("name"), it.optString("title"), it.optLong("size"), it.optString("sha1")) }
                        },
                    )
                }
            }
            return SaveListing(cards, p.optBoolean("autoSync", true))
        }
    }
}
data class SaveCard(val id: String, val name: String, val kind: String, val saves: List<SaveEntry>)
data class SaveEntry(val name: String, val title: String, val size: Long, val sha1: String)

/** A copy of one save, base64, as the host exports it. */
data class SaveFile(val cardId: String, val save: String, val title: String, val kind: String, val fileName: String, val base64: String, val sha1: String, val at: String) {
    companion object {
        fun from(p: JSONObject) = SaveFile(
            p.optString("cardId"), p.optString("save"), p.optString("title"), p.optString("kind"),
            p.optString("fileName"), p.optString("base64"), p.optString("sha1"), p.optString("at"),
        )
    }
}

/** Apollo's answer for one save: the codes that fit it. */
data class SavePatches(val cardId: String, val save: String, val gameName: String?, val productCode: String, val region: String, val attribution: List<String>, val error: String?, val codes: List<PatchCode>) {
    companion object {
        fun from(p: JSONObject): SavePatches {
            val arr = p.optJSONArray("codes")
            val attr = p.optJSONArray("attribution")
            return SavePatches(
                cardId = p.optString("cardId"), save = p.optString("save"),
                gameName = p.optString("gameName").ifEmpty { null }, productCode = p.optString("productCode"), region = p.optString("region"),
                attribution = if (attr == null) emptyList() else (0 until attr.length()).map { attr.optString(it) },
                error = p.optString("error").ifEmpty { null },
                codes = if (arr == null) emptyList() else (0 until arr.length()).mapNotNull { i ->
                    arr.optJSONObject(i)?.let { c ->
                        val opts = c.optJSONArray("options")
                        val targets = c.optJSONArray("targets")
                        PatchCode(
                            key = c.optString("key"), name = c.optString("name"), group = c.optString("group").ifEmpty { null },
                            isInfo = c.optBoolean("isInfo"), isRequired = c.optBoolean("isRequired"), isDefault = c.optBoolean("isDefault"),
                            type = c.optString("type"),
                            targets = if (targets == null) emptyList() else (0 until targets.length()).map { targets.optString(it) },
                            options = if (opts == null) emptyList() else (0 until opts.length()).mapNotNull { j ->
                                opts.optJSONObject(j)?.let { o ->
                                    val ch = o.optJSONArray("choices")
                                    PatchOption(o.optString("tag"), if (ch == null) emptyList() else (0 until ch.length()).mapNotNull { k -> ch.optJSONObject(k)?.let { it.optString("value") to it.optString("label") } })
                                }
                            },
                        )
                    }
                },
            )
        }
    }
}
data class PatchCode(val key: String, val name: String, val group: String?, val isInfo: Boolean, val isRequired: Boolean, val isDefault: Boolean, val type: String, val targets: List<String>, val options: List<PatchOption>)
data class PatchOption(val tag: String, val choices: List<Pair<String, String>>)

/** What came of an apply, push or restore. */
data class SaveResult(val cardId: String, val save: String, val ok: Boolean, val message: String, val preview: List<String>) {
    companion object {
        fun from(p: JSONObject): SaveResult {
            val pv = p.optJSONArray("preview")
            val lines = mutableListOf<String>()
            if (pv != null) for (i in 0 until pv.length()) {
                val f = pv.optJSONObject(i) ?: continue
                val first = f.optJSONArray("first")
                lines += "${f.optString("name")}: ${f.optInt("changed")} byte(s)"
                if (first != null) for (k in 0 until minOf(first.length(), 12)) {
                    val c = first.optJSONObject(k) ?: continue
                    lines += "  0x%06X: %s → %s".format(c.optInt("offset"), c.optString("from"), c.optString("to"))
                }
            }
            return SaveResult(p.optString("cardId"), p.optString("save"), p.optBoolean("ok"), p.optString("message"), lines)
        }
    }
}

enum class GhostState(val wire: String) {
    IDLE("idle"), LISTENING("listening"), THINKING("thinking"),
    SPEAKING("speaking"), SCANNING("scanning"), ERROR("error");

    companion object {
        fun from(wire: String): GhostState = entries.firstOrNull { it.wire == wire } ?: IDLE
    }
}

File: src/main/companion/protocol.ts
/**
 * The wire protocol between A-X-M and the Android companion.
 *
 * This file is the contract. It is deliberately free of Node, Electron and DOM
 * types so it can be copied verbatim into the Android project and kept in step by
 * diffing one file rather than two implementations.
 *
 * Rules that hold everywhere:
 *
 * - Every frame is one JSON object with a `protocol` number and a `type`. A peer
 *   that receives a version it does not know says so and closes, rather than
 *   guessing at a payload whose shape may have changed.
 * - `type` is a namespaced verb. The namespace is the subsystem it belongs to, so
 *   routing is a switch on the prefix and an unknown namespace is ignorable.
 * - The phone never names a file, a path or a command line. It asks for actions
 *   from a fixed list; the host decides what those mean. Nothing on this wire can
 *   express "run this program", which is the whole point.
 */

export const COMPANION_PROTOCOL_VERSION = 1;

/** Port the host listens on for discovery probes. */
export const COMPANION_DISCOVERY_PORT = 47820;
/** Default port for the control socket. The discovery reply carries the real one. */
export const COMPANION_CONTROL_PORT = 47821;
/** Magic string in a discovery probe, so we ignore unrelated broadcast traffic. */
export const COMPANION_DISCOVERY_MAGIC = "AXM-COMPANION-DISCOVER";

// ------------------------------------------------------------------ device --

/** What a phone can do. The host enables features from this, never assumes them. */
export interface CompanionCapabilities {
  touch: boolean;
  nfc: boolean;
  foldable: boolean;
  gyro: boolean;
  haptics: boolean;
  videoDecode: boolean;
  camera?: boolean;
  microphone?: boolean;
}

/** What the phone is currently showing. The host may ask it to change. */
export type CompanionMode =
  | "idle"
  | "media_remote"
  | "touchpad"
  | "xmb_controller"
  | "ds_second_screen"
  | "3ds_second_screen"
  | "toy_box"
  | "ghost";

export interface CompanionDevice {
  /** Stable per-install id the phone generates once and keeps. */
  deviceId: string;
  /** Shown in A-X-M's device list, e.g. "Pixel 9 Pro Fold". */
  name: string;
  platform: "android" | string;
  capabilities: CompanionCapabilities;
}

// ---------------------------------------------------------------- messages --

interface Frame<T extends string, P> {
  protocol: number;
  type: T;
  payload: P;
  /** Set by the sender on a request; echoed on the matching reply. */
  id?: string;
}

/** First frame on a new socket. A token means "we have paired before". */
export type HelloMessage = Frame<"device.hello", CompanionDevice & { token?: string }>;

/**
 * The host's answer to hello. `paired` false means the phone must now show the
 * code the host is displaying and send pair.submit.
 */
export type WelcomeMessage = Frame<
  "device.welcome",
  { hostName: string; hostVersion: string; paired: boolean; protocol: number }
>;

/** The phone sends back the code the user read off the A-X-M screen. */
export type PairSubmitMessage = Frame<"pair.submit", { code: string }>;

/** On success the phone stores the token and skips the code next time. */
export type PairResultMessage = Frame<
  "pair.result",
  { ok: boolean; token?: string; reason?: string; attemptsLeft?: number }
>;

export type GoodbyeMessage = Frame<"device.goodbye", { reason?: string }>;

/** Keeps NAT and sleepy Wi-Fi from quietly dropping an idle socket. */
export type PingMessage = Frame<"link.ping", { at: number }>;
export type PongMessage = Frame<"link.pong", { at: number }>;

/** Host tells the phone what screen to show; the phone may also request one. */
export type ModeMessage = Frame<"companion.mode", { mode: CompanionMode; platform?: string }>;

/** Menu navigation. Exactly the actions the menu already understands. */
export type XmbAction = "up" | "down" | "left" | "right" | "confirm" | "back" | "context" | "guide";
export type XmbMessage = Frame<"xmb.input", { action: XmbAction }>;

export type MediaCommand =
  | "play"
  | "pause"
  | "toggle"
  | "next"
  | "previous"
  | "stop"
  | "volume"
  | "mute"
  | "shuffle"
  | "repeat"
  | "favorite"
  | "seek"
  /** Absolute position in seconds - the phone dragging the progress bar. */
  | "position"
  /** value 1: the phone plays the audio from here on (the menu pauses); 0: back to the menu. */
  | "route";
export type MediaCommandMessage = Frame<"media.command", { command: MediaCommand; value?: number }>;

/** Host is the authority on what is playing; the phone mirrors this. */
export type MediaStateMessage = Frame<
  "media.state",
  {
    playing: boolean;
    title?: string;
    artist?: string;
    album?: string;
    /** http URL on the host's own asset endpoint, never a local file path. */
    artworkUrl?: string;
    positionSeconds?: number;
    durationSeconds?: number;
    volume?: number;
    muted?: boolean;
    /** The file itself, streamed from the menu, for playing on the phone. */
    streamUrl?: string;
    /** "host" or "phone". */
    output?: string;
  }
>;

/** Touchpad. Deltas, not absolute screen positions, so DPI never has to match. */
export type PointerMessage = Frame<
  "pointer.input",
  {
    kind: "move" | "click" | "scroll" | "down" | "up";
    dx?: number;
    dy?: number;
    button?: "left" | "right" | "middle";
  }
>;

/** A tag the phone read. Identification only - never tag contents or keys. */
export type ToyScanMessage = Frame<
  "toybox.scan",
  {
    uid: string;
    technology?: string;
    /** Amiibo head/tail, or a character/variant pair, whichever the tag gives. */
    head?: string;
    tail?: string;
    characterId?: string;
    variantId?: string;
  }
>;

/** What the host made of that tag, plus which installed games it fits. */
export type ToyMetadataMessage = Frame<
  "toybox.metadata",
  {
    figureId: string | null;
    name: string;
    franchise?: string;
    series?: string;
    variant?: string;
    artworkUrl?: string;
    compatibleGames: { id: string; name: string; installed: boolean }[];
  }
>;

export type GhostState = "idle" | "listening" | "thinking" | "speaking" | "scanning" | "error";
export type GhostStateMessage = Frame<"ghost.state", { state: GhostState }>;
export type GhostMessageFrame = Frame<"ghost.message", { text: string; spoken?: boolean }>;

// ---------------------------------------------------------- settings --

/**
 * A menu setting the phone may change. The list is the host's: it says what each
 * one is called, what it can be, and what it is now. The phone only ever sends
 * back an id and one of the values it was given.
 */
export interface CompanionSetting {
  id: string;
  title: string;
  /** "Audio", "Display", "TV Streaming"... - the phone groups by this. */
  group: string;
  kind: "toggle" | "choice";
  /** "on" / "off" for a toggle; an option id for a choice. */
  value: string;
  options?: { id: string; label: string }[];
  /** One line under the title, as the menu shows it. */
  detail?: string;
}
export type SettingsStateMessage = Frame<"settings.state", { items: CompanionSetting[] }>;
export type SettingsSetMessage = Frame<"settings.set", { id: string; value: string }>;

// ---------------------------------------------------------- keyboard --

/** The menu is asking for text; the phone can type it with its own keyboard. */
export type KeyboardShowMessage = Frame<"keyboard.show", { title: string; label: string; value: string; secret: boolean }>;
export type KeyboardHideMessage = Frame<"keyboard.hide", { at: number }>;
/** What the phone typed so far; done = the user pressed enter / next. */
export type KeyboardInputMessage = Frame<"keyboard.input", { text: string; done: boolean }>;

// ------------------------------------------------------------- music --

/**
 * Browsing the music library from the phone. Folders and files are named by
 * opaque keys the host hands out in a listing; the phone never sends a path.
 * No key means the top level: every drive's MUSIC folder and the user's own.
 */
export type MusicBrowseMessage = Frame<"music.browse", { key?: string }>;
export type MusicListingMessage = Frame<
  "music.listing",
  { key: string; name: string; parent?: string; entries: { key: string; name: string; kind: "folder" | "file" }[] }
>;
export type MusicPlayMessage = Frame<"music.play", { key: string }>;

// --------------------------------------------------------------- saves ----

/**
 * Memory card saves on the phone: a copy of each save the menu manages, kept
 * offline, and Apollo's cheats run on the host at the phone's request. The
 * phone names saves by the card id and save name the host listed; the bytes
 * travel base64 in both directions. The host is the only thing that touches a
 * card, so an edit from the phone and one from the TV are the same code path.
 */
export interface SaveListing {
  cards: { id: string; name: string; kind: "ps1" | "ps2"; saves: { name: string; title: string; size: number; sha1: string }[] }[];
  /** True while the host will push copies on its own (Settings › Companion). */
  autoSync: boolean;
}
export type SavesListMessage = Frame<"saves.list", SaveListing>;
/** A copy of one save: .mcs blocks (PS1) or .psu (PS2), base64. */
export type SavesFileMessage = Frame<"saves.file", { cardId: string; save: string; title: string; kind: "ps1" | "ps2"; fileName: string; base64: string; sha1: string; at: string }>;
export type SavesRequestMessage = Frame<"saves.request", { cardId: string; save: string }>;
/** The phone's copy back onto the card, replacing what's there (the card is backed up first). */
export type SavesPushMessage = Frame<"saves.push", { cardId: string; save: string; base64: string }>;
export type SavesCheatsMessage = Frame<"saves.cheats", { cardId: string; save: string }>;
export type SavesPatchesMessage = Frame<
  "saves.patches",
  {
    cardId: string;
    save: string;
    gameName: string | null;
    productCode: string;
    region: string;
    attribution: string[];
    error?: string;
    codes: { key: string; name: string; group: string | null; isInfo: boolean; isRequired: boolean; isDefault: boolean; type: string; targets: string[]; options: { tag: string; choices: { value: string; label: string }[] }[] }[];
  }
>;
export type SavesApplyMessage = Frame<"saves.apply", { cardId: string; save: string; selections: { key: string; options: Record<string, string> }[]; preview?: boolean }>;
export type SavesRestoreMessage = Frame<"saves.restore", { cardId: string; save: string }>;
/** What came of an apply / push / restore, in words, plus a preview when one was asked for. */
export type SavesResultMessage = Frame<"saves.result", { cardId: string; save: string; ok: boolean; message: string; preview?: { name: string; changed: number; first: { offset: number; from: string; to: string }[] }[] }>;

// --------------------------------------------------------------- phone ----

/**
 * Experimental: the phone's own Phone, Messages and Contacts, used from the menu's
 * Device column. The phone is the one that owns the data and the SIM, so the
 * direction flips here: the host asks, and the phone decides - it answers only
 * while the user has switched sharing on in the phone app and granted Android's
 * permissions, and it refuses anything else. Numbers and message text are the
 * only free-form values, and they only ever go to the phone's own dialer and SMS.
 */
export interface PhoneShareState {
  /** The user switched each on in the phone app, and Android granted it. */
  calls: boolean;
  messages: boolean;
  contacts: boolean;
  /** False on a Wi-Fi-only tablet: contacts work, calls and texts cannot. */
  telephony: boolean;
  /** "phone" or "tablet" - the phone app's own idea of what it is. */
  formFactor: "phone" | "tablet";
}
export type PhoneStateMessage = Frame<"phone.state", PhoneShareState>;

export interface PhoneContact {
  id: string;
  name: string;
  numbers: { number: string; label: string }[];
  starred: boolean;
  /** A small JPEG as a data: URL, when the contact has a picture. */
  photo?: string;
}
export interface PhoneThread {
  id: string;
  address: string;
  name?: string;
  snippet: string;
  /** Epoch milliseconds. */
  date: number;
  unread: number;
}
export interface PhoneSms {
  id: string;
  body: string;
  date: number;
  outgoing: boolean;
}
export interface PhoneCall {
  number: string;
  name?: string;
  date: number;
  durationSeconds: number;
  kind: "incoming" | "outgoing" | "missed" | "rejected";
}

/** Host asks for a list; the phone answers with the matching message below. */
export type PhoneRequestMessage = Frame<"phone.request", { what: "contacts" | "threads" | "calls" | "messages"; threadId?: string }>;
export type PhoneContactsMessage = Frame<"phone.contacts", { contacts: PhoneContact[] }>;
export type PhoneThreadsMessage = Frame<"phone.threads", { threads: PhoneThread[] }>;
export type PhoneMessagesMessage = Frame<"phone.messages", { threadId: string; address: string; messages: PhoneSms[] }>;
export type PhoneCallsMessage = Frame<"phone.calls", { calls: PhoneCall[] }>;

/** Place a call, answer the ringing one, or end the current one. */
export type PhoneDialMessage = Frame<"phone.dial", { number: string }>;
export type PhoneAnswerMessage = Frame<"phone.answer", { at: number }>;
export type PhoneHangupMessage = Frame<"phone.hangup", { at: number }>;
/** The phone's call, as it changes: ringing (with who), offhook (a call is up), idle. */
export type PhoneCallStateMessage = Frame<"phone.callState", { state: "idle" | "ringing" | "offhook"; number?: string; name?: string }>;

/** Send a text. The phone replies with phone.smsResult carrying the same ref. */
export type PhoneSendSmsMessage = Frame<"phone.sendSms", { to: string; body: string; ref: string }>;
export type PhoneSmsResultMessage = Frame<"phone.smsResult", { ref: string; ok: boolean; error?: string }>;
/** A text arrived on the phone. */
export type PhoneIncomingSmsMessage = Frame<"phone.incomingSms", { threadId: string; address: string; name?: string; body: string; date: number }>;

/** Anything the host wants to say went wrong, in words fit to show a user. */
export type ErrorMessage = Frame<"link.error", { message: string; fatal?: boolean }>;

export type CompanionMessage =
  | HelloMessage
  | WelcomeMessage
  | PairSubmitMessage
  | PairResultMessage
  | GoodbyeMessage
  | PingMessage
  | PongMessage
  | ModeMessage
  | XmbMessage
  | MediaCommandMessage
  | MediaStateMessage
  | PointerMessage
  | ToyScanMessage
  | ToyMetadataMessage
  | GhostStateMessage
  | GhostMessageFrame
  | SettingsStateMessage
  | SettingsSetMessage
  | KeyboardShowMessage
  | KeyboardHideMessage
  | KeyboardInputMessage
  | MusicBrowseMessage
  | MusicListingMessage
  | MusicPlayMessage
  | SavesListMessage
  | SavesFileMessage
  | SavesRequestMessage
  | SavesPushMessage
  | SavesCheatsMessage
  | SavesPatchesMessage
  | SavesApplyMessage
  | SavesRestoreMessage
  | SavesResultMessage
  | PhoneStateMessage
  | PhoneRequestMessage
  | PhoneContactsMessage
  | PhoneThreadsMessage
  | PhoneMessagesMessage
  | PhoneCallsMessage
  | PhoneDialMessage
  | PhoneAnswerMessage
  | PhoneHangupMessage
  | PhoneCallStateMessage
  | PhoneSendSmsMessage
  | PhoneSmsResultMessage
  | PhoneIncomingSmsMessage
  | ErrorMessage;

/** What the phone sends about its Phone, Messages and Contacts, forwarded to the menu. */
export const PHONE_INBOUND_TYPES = [
  "phone.state",
  "phone.contacts",
  "phone.threads",
  "phone.messages",
  "phone.calls",
  "phone.callState",
  "phone.smsResult",
  "phone.incomingSms",
] as const;
export type PhoneInboundType = (typeof PHONE_INBOUND_TYPES)[number];

export type CompanionMessageType = CompanionMessage["type"];

/**
 * Messages a phone is allowed to send *before* it has paired. Everything else is
 * refused until the code has been accepted, so an unpaired device on the network
 * can do nothing but introduce itself.
 */
export const PRE_PAIR_TYPES: CompanionMessageType[] = ["device.hello", "pair.submit", "link.ping", "device.goodbye"];

export function frame<T extends CompanionMessage>(type: T["type"], payload: T["payload"], id?: string): string {
  return JSON.stringify({ protocol: COMPANION_PROTOCOL_VERSION, type, payload, ...(id ? { id } : {}) });
}

/**
 * Parses a frame, returning null for anything malformed. Never throws: this runs on
 * bytes from the network, and a bad frame must not take the server down.
 */
export function parseFrame(raw: string): CompanionMessage | null {
  try {
    const value = JSON.parse(raw) as Partial<CompanionMessage>;
    if (!value || typeof value !== "object") return null;
    if (typeof value.type !== "string") return null;
    if (typeof value.protocol !== "number") return null;
    if (value.payload === undefined || value.payload === null) return null;
    return value as CompanionMessage;
  } catch {
    return null;
  }
}
